CPP/DesktopToastsCppApp/DesktopToastsSample.cpp

```diff
@@ -0,0 +1,552 @@
+#pragma comment(linker,"\"/manifestdependency:type='win32' name='Microsoft.Windows.Common-Controls' version='6.0.0.0' processorArchitecture='*' publicKeyToken='6595b64144ccf1df' language='*'\"")
+
+#include <atlstr.h>
+#include <SDKDDKVer.h>
+#include <Windows.h>
+#include <Psapi.h>
+#include <string>
+#include <strsafe.h>
+#include <ShObjIdl.h>
+#include <Shlobj.h>
+#include <Pathcch.h>
+#include <propvarutil.h>
+#include <propkey.h>
+#include <wrl.h>
+#include <wrl\wrappers\corewrappers.h>
+#include <windows.ui.notifications.h>
+#include "NotificationActivationCallback.h"
+#include "DesktopNotificationManagerCompat.h"
+
+#define THROW_IF_FAILED(hr) if (FAILED(hr)) { throw hr; }
```

Comment:
1. Just throw std::exception instead of throwing an HRESULT.
2. This code is actually does a terrible thing. Don't feel bad, this is a common problem when writing this kind of macro. If we expand this out...

THROW_IF_FAILED(m_history->Remove(HStringReference(tag).Get()));  =>
if (FAILED(m_history->Remove(HStringReference(tag).Get()))) { throw m_history->Remove(HStringReference(tag).Get()); }

instead, let's do this:

```
#define THROW_IF_FAILED(hr)                                             \
do {                                                                    \
    HRESULT _hrTemp = hr;                                               \
    if (FAILED(_hrTemp))                                                \
    {                                                                   \
        std::error_code _errCode(_hrTemp,std::generic_category());      \
        throw std::system_error(_errCode);                              \
    }                                                                   \
} while(false)
```